StateTransferTracker#cacheTopologyUpdated blocking issue

### Describe the bug

I have seen it pop up a few times in various PRs but there is a new blocking issue:

```
java.lang.AssertionError: Blocking call! jdk.internal.misc.Unsafe#park on thread Thread[#6790,non-blocking-thread-ServerResourceTest-NodeB-54-4,5,ISPN-non-blocking-group]
	at org.infinispan.util.CoreTestBlockHoundIntegration.lambda$applyTo$0(CoreTestBlockHoundIntegration.java:51)
	at reactor.blockhound.BlockHound$Builder.lambda$install$8(BlockHound.java:488)
	at reactor.blockhound.BlockHoundRuntime.checkBlocking(BlockHoundRuntime.java:89)
	at java.base/jdk.internal.misc.Unsafe.park(Unsafe.java)
	at java.base/java.util.concurrent.locks.LockSupport.park(LockSupport.java:223)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:790)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1030)
	at java.base/java.util.concurrent.locks.ReentrantLock$Sync.lock(ReentrantLock.java:154)
	at java.base/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:323)
	at org.infinispan.statetransfer.StateTransferTracker.cacheTopologyUpdated(StateTransferTracker.java:149)
	at org.infinispan.statetransfer.StateTransferManagerImpl$1.onTopologyReceived(StateTransferManagerImpl.java:125)
	at org.infinispan.topology.LocalCacheStatus.setCurrentTopology(LocalTopologyManagerImpl.java:1028)
	at org.infinispan.topology.LocalTopologyManagerImpl.updateCacheTopology(LocalTopologyManagerImpl.java:566)
	at org.infinispan.topology.LocalTopologyManagerImpl.doHandleTopologyUpdate(LocalTopologyManagerImpl.java:488)
	at org.infinispan.topology.LocalTopologyManagerImpl.lambda$handleTopologyUpdate$1(LocalTopologyManagerImpl.java:448)
	at org.infinispan.topology.LocalTopologyManagerImpl.lambda$orderOnCache$0(LocalTopologyManagerImpl.java:953)
	at org.infinispan.util.concurrent.ActionSequencer.safeNonBlockingCall(ActionSequencer.java:59)
	at org.infinispan.util.concurrent.ActionSequencer$SequenceEntry.run(ActionSequencer.java:228)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:405)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1474)
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

--- a/core/src/main/java/org/infinispan/statetransfer/StateTransferTracker.java
+++ b/core/src/main/java/org/infinispan/statetransfer/StateTransferTracker.java
@@ -67,6 +67,16 @@ public class StateTransferTracker {
 
    private final List<Entry> listeners = new CopyOnWriteArrayList<>();
 
+   // This method is here solely for Blockhound, do not move/modify unless changing CoreBlockHoundIntegration
+   private void acquireLock() {
+      lock.lock();
+   }
+
+   // This method isn't really needed, but here for symmetry with acquireLock
+   private void releaseLock() {
+      lock.unlock();
+   }
+
    /**
     * Checks if a state transfer is currently in progress.
     *
@@ -78,11 +88,11 @@ public class StateTransferTracker {
     * @return {@code true} if state transfer is active; {@code false} otherwise.
     */
    public boolean isStateTransferInProgress() {
-      lock.lock();
+      acquireLock();
       try {
          return stable.isPending();
       } finally {
-         lock.unlock();
+         releaseLock();
       }
    }
 
@@ -111,7 +121,7 @@ public class StateTransferTracker {
     * @return A {@link CompletionStage} that completes when the listener returns {@code true}.
     */
    public CompletionStage<Void> onStateTransferCompleted(BiPredicate<CacheTopology, Throwable> listener) {
-      lock.lock();
+      acquireLock();
       try {
          if (log.isTraceEnabled())
             log.tracef("waiting state completion %s (consumer=%s) (provider=%s)%n", cacheName, consumer.isPending(), provider.isPending());
@@ -129,7 +139,7 @@ public class StateTransferTracker {
          listeners.add(new Entry(cf, listener));
          return cf;
       } finally {
-         lock.unlock();
+         releaseLock();
       }
    }
 
@@ -146,7 +156,7 @@ public class StateTransferTracker {
    public void cacheTopologyUpdated(CacheTopology cacheTopology) {
       boolean isStableTopology = cacheTopology.getPendingCH() == null;
       if (isStableTopology) {
-         lock.lock();
+         acquireLock();
          try {
             log.tracef("Installed stable topology for %s with %s, state transfer is done now", cacheName, cacheTopology);
             int previousTopologyId = currentTopology != null ? currentTopology.getTopologyId() : Integer.MIN_VALUE;
@@ -166,7 +176,7 @@ public class StateTransferTracker {
             consumer.complete(previousTopologyId, null);
             provider.complete(currentTopology.getTopologyId(), null);
          } finally {
-            lock.unlock();
+            releaseLock();
          }
       }
    }
@@ -182,12 +192,12 @@ public class StateTransferTracker {
     * @param topologyId The topology ID associated with this state transfer start.
     */
    public void startStateConsumer(int topologyId) {
-      lock.lock();
+      acquireLock();
       try {
          log.tracef("starting state consumer %s", cacheName);
          checkTopologyId(topologyId);
       } finally {
-         lock.unlock();
+         releaseLock();
       }
    }
 
@@ -197,12 +207,12 @@ public class StateTransferTracker {
     * @param topologyId The topology ID that completed.
     */
    public void completeStateConsumer(int topologyId) {
-      lock.lock();
+      acquireLock();
       try {
          log.tracef("stopping state consumer %s", cacheName);
          consumer.complete(topologyId, null);
       } finally {
-        lock.unlock();
+        releaseLock();
       }
    }
 
@@ -217,12 +227,12 @@ public class StateTransferTracker {
     * @param topologyId The topology ID associated with this state transfer start.
     */
    public void startStateProvider(int topologyId) {
-      lock.lock();
+      acquireLock();
       try {
          log.tracef("starting state provider %s", cacheName);
          checkTopologyId(topologyId);
       } finally {
-         lock.unlock();
+         releaseLock();
       }
    }
 
@@ -232,12 +242,12 @@ public class StateTransferTracker {
     * @param topologyId The topology ID that completed.
     */
    public void completeStateProvider(int topologyId) {
-      lock.lock();
+      acquireLock();
       try {
          log.tracef("stopping state provider %s", cacheName);
          provider.complete(topologyId, null);
       } finally {
-         lock.unlock();
+         releaseLock();
       }
    }
 

--- a/core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java
+++ b/core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java
@@ -18,6 +18,7 @@ import org.infinispan.persistence.sifs.TemporaryTable;
 import org.infinispan.remoting.transport.jgroups.JGroupsTransport;
 import org.infinispan.statetransfer.StateConsumerImpl;
 import org.infinispan.statetransfer.StateTransferLockImpl;
+import org.infinispan.statetransfer.StateTransferTracker;
 import org.infinispan.topology.ClusterCacheStatus;
 import org.infinispan.topology.ClusterTopologyManagerImpl;
 import org.infinispan.topology.LocalTopologyManagerImpl;
@@ -74,6 +75,9 @@ public class CoreBlockHoundIntegration implements BlockHoundIntegration {
          builder.allowBlockingCallsInside(PersistenceManagerImpl.class.getName(), "acquireReadLock");
 
          builder.allowBlockingCallsInside(JGroupsTransport.class.getName(), "withView");
+
+         // StateTransferTracker holds the lock briefly to update topology state and complete futures
+         builder.allowBlockingCallsInside(StateTransferTracker.class.getName(), "acquireLock");
       }
       // This invokes the actual runnable - we have to make sure it doesn't block as normal
       builder.disallowBlockingCallsInside(LimitedExecutor.class.getName(), "actualRun");

